satellite/metainfo/metainfo.go

```diff
@@ -1414,6 +1475,18 @@ func (endpoint *Endpoint) CommitSegment(ctx context.Context, req *pb.SegmentComm
 		return nil, rpcstatus.Error(rpcstatus.Unauthenticated, err.Error())
 	}
 
+	if numResults := len(req.UploadResult); numResults < int(streamID.Redundancy.GetSuccessThreshold()) {
+		endpoint.log.Debug("the results of uploaded pieces for the segment is below the redundancy optimal threshold",
+			zap.Int("upload pieces results", numResults),
+			zap.Int32("redundancy optimal threshold", streamID.Redundancy.GetSuccessThreshold()),
+			zap.Stringer("segment ID", req.SegmentId),
+		)
+		return nil, rpcstatus.Errorf(rpcstatus.InvalidArgument,
```

Comment:
if this is a new case where an error will be returned where it wasn't previously, does a test need to be added?

Reply:
I spent quite a few time to find out is possible to test it.

Mostly, I thought that there was no test for this kind of thing because this is not a new returned error, it is mostly a premature check that can stop processing at this point rather than doing more things to catch it.

The check was already done by the call to `filterValidPieces` (in the [line 1535 of this file]https://github.com/storj/storj/pull/3218/files/bd98ae7351f2853074e0f45669f98ac20f948773#diff-ea4e4260da09e3a291a61fd90bd2d656R1535); such function does the check at the end of it ([line 570 of this file](https://github.com/storj/storj/pull/3218/files/bd98ae7351f2853074e0f45669f98ac20f948773#diff-ea4e4260da09e3a291a61fd90bd2d656R570) but it performs the check after discarding invalid pieces, getting the same result if the number of sent pieces is already lower than the optimal threshold. Hence it isn't new, it's mostly an improvement to return the error in advance and avoid further computation.

Thanks to your comment, I found out that's possible to test and there is already a test which is checking it https://github.com/storj/storj/blob/master/satellite/metainfo/metainfo_test.go#L298

There is no need for a new test, mostly it has to pass it as it was, however, I think that's worth to check the gRPC status code and that it's what I'm going to add.

Reply:
Changes in test done!

Reply:
thanks a lot for looking into this in detail Ivan!